cpp/include/cuvs/neighbors/scann.hpp

```diff
@@ -0,0 +1,310 @@
+
+/**
+ * @defgroup scann_cpp_index ScaNN index type
+ * @{
+ */
+
+/**
+ * @brief ScaNN index.
+ *
+ * The index stores the dataset and the ScaNN graph in device memory.
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the vector indices (represent dataset.extent(0))
+ *
+ */
+template <typename T, typename IdxT>
+struct index : cuvs::neighbors::index {
+  static_assert(!raft::is_narrowing_v<uint32_t, IdxT>,
+                "IdxT must be able to represent all values of uint32_t");
+
+ public:
+  /** Distance metric used for clustering. */
+  [[nodiscard]] constexpr inline auto metric() const noexcept -> cuvs::distance::DistanceType
+  {
+    return metric_;
+  }
+
+  /** Total length of the index (number of vectors). */
+  IdxT size() const noexcept;
+
+  /** Dimensionality of the data. */
+  [[nodiscard]] constexpr inline auto dim() const noexcept -> uint32_t { return dim_; }
+
+  // Don't allow copying the index for performance reasons (try avoiding copying data)
+  index(const index&)                    = delete;
+  index(index&&)                         = default;
+  auto operator=(const index&) -> index& = delete;
+  auto operator=(index&&) -> index&      = default;
+  ~index()                               = default;
+
+  /** Construct an empty index. It will need to be trained and populated with `build`*/
+  //  index(raft::resources const& res) {}
+
+  index(raft::resources const& res,
+        cuvs::distance::DistanceType metric,
+        uint32_t scann_clusters,
+        uint32_t pq_bits,
+        uint32_t pq_dim,
+        IdxT n_rows,
+        IdxT dim,
+        uint32_t pq_clusters,
+        uint32_t pq_num_subspaces,
+        bool bf16_enabled)
+    : cuvs::neighbors::index(),
+      metric_(metric),
+      pq_dim_(pq_dim),
+      pq_bits_(pq_bits),
+      scann_clusters_(scann_clusters),
+      centers_(raft::make_device_matrix<float, IdxT>(res, scann_clusters, dim)),
+      labels_(raft::make_device_vector<uint32_t, IdxT>(res, n_rows)),
+      soar_labels_(raft::make_device_vector<uint32_t, IdxT>(res, n_rows)),
+      pq_codebook_(
+        raft::make_device_matrix<float, uint32_t, raft::row_major>(res, pq_clusters, dim)),
+      quantized_residuals_(
+        raft::make_host_matrix<uint8_t, IdxT, raft::row_major>(n_rows, pq_num_subspaces)),
+      quantized_soar_residuals_(
+        raft::make_host_matrix<uint8_t, IdxT, raft::row_major>(n_rows, pq_num_subspaces)),
+      n_rows_(n_rows),
+      dim_(dim),
+      bf16_dataset_(raft::make_host_matrix<int16_t, IdxT, raft::row_major>(
+        bf16_enabled ? n_rows : 0, bf16_enabled ? dim : 0))
+
+  {
+  }
+
+  index(raft::resources const& res, const index_params& params, IdxT n_rows, IdxT dim)
+    : index(res,
+            params.metric,
+            params.scann_clusters,
+            params.pq_bits,
+            params.pq_dim,
+            n_rows,
+            dim,
+            1 << params.pq_bits,
+            dim / params.pq_dim,
```

Comment:
If user sets `pq_dim > dim`, that would zero `pq_num_subspaces`. I assume that is not valid. Please add checks (see usage of `RAFT_EXPECTS`).